addon/-private/system/relationships/state/implicit.js

```diff
@@ -0,0 +1,209 @@
+
+  addInternalModels(internalModels, idx) {
+    heimdall.increment(addInternalModels);
+    for (let i=0; i < internalModels.length; i++) {
+      this.addInternalModel(internalModels[i], idx);
+
+      if (idx !== undefined) {
+        idx++;
+      }
+    }
+  }
+
+  addCanonicalInternalModels(records, idx) {
+    heimdall.increment(addCanonicalInternalModels);
+    for (let i=0; i<records.length; i++) {
+      this.addCanonicalInternalModel(records[i], i+idx);
+      if (idx !== undefined) {
+        idx++;
+      }
+    }
+  }
+
+  addCanonicalInternalModel(record) {
+    heimdall.increment(addCanonicalInternalModel);
+    if (!this.canonicalState.has(record)) {
+      this.canonicalState.add(record);
+    }
+    this.flushCanonicalLater();
+    this.setHasData(true);
+  }
+
+  removeCanonicalInternalModels(records, idx) {
+    heimdall.increment(removeCanonicalInternalModels);
+    for (let i=0; i<records.length; i++) {
+      if (idx !== undefined) {
+        this.removeCanonicalInternalModel(records[i], i+idx);
+      } else {
+        this.removeCanonicalInternalModel(records[i]);
+      }
+    }
+  }
+
+  removeCanonicalInternalModel(record, idx) {
+    heimdall.increment(removeCanonicalInternalModel);
+    if (this.canonicalState.has(record)) {
+      this.removeCanonicalInternalModelFromOwn(record);
+      if (this.inverseKey) {
+        this.removeCanonicalInternalModelFromInverse(record);
+      }
+    }
+    this.flushCanonicalLater();
+  }
+
+  addInternalModel(record, idx) {
+    heimdall.increment(addInternalModel);
+    if (!this.currentState.has(record)) {
+      this.currentState.addWithIndex(record, idx);
+      this.notifyRecordRelationshipAdded(record, idx);
+      this.internalModel.updateRecordArrays();
+    }
+    this.setHasData(true);
+  }
+
+  removeInternalModel(record) {
+    heimdall.increment(removeInternalModel);
+    if (this.currentState.has(record)) {
+      this.removeInternalModelFromOwn(record);
+      if (this.inverseKey) {
+        this.removeInternalModelFromInverse(record);
+      }
+    }
+  }
+
+  removeInternalModelFromOwn(record) {
+    heimdall.increment(removeInternalModelFromOwn);
+    this.currentState.delete(record);
+    this.notifyRecordRelationshipRemoved(record);
+    this.internalModel.updateRecordArrays();
+  }
+
+  removeCanonicalInternalModelFromOwn(record) {
+    heimdall.increment(removeCanonicalInternalModelFromOwn);
+    this.canonicalState.delete(record);
+    this.flushCanonicalLater();
+  }
+
+  flushCanonical() {
+    heimdall.increment(flushCanonical);
+    let list = this.currentState.list;
+    this.willSync = false;
+    //a hack for not removing new records
+    //TODO remove once we have proper diffing
+    let newInternalModels = [];
+    for (let i = 0; i < list.length; i++) {
+      if (list[i].isNew()) {
+        newInternalModels.push(list[i]);
+      }
+    }
+
+    //TODO(Igor) make this less abysmally slow
+    this._currentState = this.canonicalState.copy();
```

Comment:
it doesn't matter here in theory but I did mean to change this to use it once I added the setter.